includes/gateway/class-omise-payment-creditcard.php

```diff
@@ -166,235 +169,192 @@ public function payment_fields() {
+			$data    = array(
+				'amount'      => $this->format_amount_subunit( $order->get_total(), $order->get_order_currency() ),
+				'currency'    => $order->get_order_currency(),
+				'description' => apply_filters( 'omise_charge_params_description', 'WooCommerce Order id ' . $order_id, $order ),
+				'return_uri'  => add_query_arg( 'order_id', $order_id, site_url() . '?wc-api=omise_callback' )
+			);
 
-				$this->set_order_transaction_id( $charge['id'] );
+			if ( ! empty( $omise_customer_id ) && ! empty( $card_id ) ) {
+				$data['customer'] = $omise_customer_id;
+				$data['card']     = $card_id;
+			} else {
+				$data['card'] = $token;
+			}
 
-				if ( Omise_Charge::is_failed( $charge ) ) {
-					throw new Exception( Omise_Charge::get_error_message( $charge ) );
-				}
+			// Set capture status (otherwise, use API's default behaviour)
+			if ( self::PAYMENT_ACTION_AUTHORIZE_CAPTURE === $this->payment_action ) {
+				$data['capture'] = true;
+			} else if ( self::PAYMENT_ACTION_AUTHORIZE === $this->payment_action ) {
+				$data['capture'] = false;
+			}
+			$metadata = apply_filters( 'omise_charge_params_metadata', array(), $order );
 
-				if ( $this->omise_3ds ) {
-					$order->add_order_note(
-						sprintf(
-							__( 'Omise: Processing a 3-D Secure payment, redirecting buyer to %s', 'omise' ),
-							esc_url( $charge['authorize_uri'] )
-						)
-					);
+			$data['metadata'] = array_merge( $metadata, array(
+				/** override order_id as a reference for webhook handlers **/
+				/** backward compatible with WooCommerce v2.x series **/
+				'order_id' => version_compare( WC()->version, '3.0.0', '>=' ) ? $order->get_id() : $order->id
+			) );
 
-					return array(
-						'result'   => 'success',
-						'redirect' => $charge['authorize_uri'],
-					);
-				} else {
-					switch ( strtoupper( $this->payment_action ) ) {
-						case 'MANUAL_CAPTURE':
-							$success = Omise_Charge::is_authorized( $charge );
-							if ( $success ) {
-								$order->add_order_note(
-									sprintf(
-										wp_kses(
-											__( 'Omise: Payment processing.<br/>An amount of %1$s %2$s has been authorized', 'omise' ),
-											array( 'br' => array() )
-										),
-										$order->get_total(),
-										$order->get_order_currency()
-									)
-								);
-							}
-
-							break;
-
-						case 'AUTO_CAPTURE':
-							$success = Omise_Charge::is_paid( $charge );
-							if ( $success ) {
-								$order->payment_complete();
-								$order->add_order_note(
-									sprintf(
-										wp_kses(
-											__( 'Omise: Payment successful.<br/>An amount of %1$s %2$s has been paid', 'omise' ),
-											array( 'br' => array() )
-										),
-										$order->get_total(),
-										$order->get_order_currency()
-									)
-								);
-							}
+			return OmiseCharge::create( $data );
+		}
 
-							break;
+		/**
+		 * @inheritdoc
+		 */
+		public function result( $order_id, $order, $charge ) {
+			if ( Omise_Charge::is_failed( $charge ) ) {
```

Comment:
Sometimes `Omise_Charge::is_failed` is used, sometimes just a string check on charge status - should these all be standardised?

Reply:
👍 Agreed.
Actually, I have a plan to remove `Omise_Charge` and `includes/libraries/omise-plugin` later on (as it might not really related to this refactoring). Guess it will be more easy to review.